src/engine/SortedUnionImpl.h

```diff
@@ -0,0 +1,241 @@
+      }
+    }
+    return false;
+  }
+
+  // Fetch the next element from the range, make a copy if it's a `Wrapper`.
+  // Otherwise move it. If the range is exhausted, return `std::nullopt`.
+  std::optional<Result::IdTableVocabPair> passNext(
+      const std::vector<ColumnIndex>& permutation, auto& it, auto end,
+      size_t& index) {
+    if (it != end) {
+      Result::IdTableVocabPair pair = moveOrCopy(*it);
+      pair.idTable_ = applyPermutation_(std::move(pair.idTable_), permutation);
+
+      index = 0;
+      ++it;
+      return pair;
+    }
+    return std::nullopt;
+  }
+
+  // Append the remainder of the last partially consumed table to the current
+  // result table and merge the local vocabs.
+  void appendCurrent(const std::vector<ColumnIndex>& permutation, auto& it,
+                     size_t& index) {
+    resultTable_.insertAtEnd(it->idTable_, index, std::nullopt, permutation,
+                             Id::makeUndefined());
+    localVocab_.mergeWith(std::span{&it->localVocab_, 1});
+    index = 0;
+    ++it;
+  }
+
+  // For the non-lazy case just append the remaining tables to the aggregated
+  // result table until the range is exhausted.
+  void appendRemaining(const std::vector<ColumnIndex>& permutation, auto& it,
+                       auto end, size_t& index) {
+    while (it != end) {
+      appendCurrent(permutation, it, index);
+    }
+  }
+
+  // Write a new row to the result table. The parameter `left` controls which
+  // permutation is used to write the row to the result table.
+  void pushRow(bool left, const auto& row) {
+    resultTable_.emplace_back();
+    for (size_t column = 0; column < resultTable_.numColumns(); column++) {
+      ColumnIndex origin = columnOrigins_.at(column).at(!left);
+      resultTable_.at(resultTable_.size() - 1, column) =
+          origin == Union::NO_COLUMN ? Id::makeUndefined() : row[origin];
+    }
+  }
+
+  // Increment the iterators if the current id table is fully processed and
+  // reset the index back to zero.
+  void advanceRangeIfConsumed() {
+    if (index1_ == it1_.value()->idTable_.size()) {
+      ++it1_.value();
+      index1_ = 0;
+    }
+    if (index2_ == it2_.value()->idTable_.size()) {
+      ++it2_.value();
+      index2_ = 0;
+    }
+  }
+
+  // Retrieve the current result from `resultTable_` and `localVocab_` and reset
+  // those members back to their initial value so the next operation can
+  // continue adding values.
+  Result::IdTableVocabPair popResult() {
+    auto result = Result::IdTableVocabPair{std::move(resultTable_),
+                                           std::move(localVocab_)};
+    resultTable_ = IdTable{resultTable_.numColumns(), allocator_};
+    resultTable_.reserve(Union::chunkSize);
+    localVocab_ = LocalVocab{};
+    return result;
+  }
+
+  // ___________________________________________________________________________
+  std::optional<Result::IdTableVocabPair> get() override {
+    if (aggregatedTableReturned_) {
```

Comment:
Can you rename  this to ` isCompletelyFinished_`?

Reply:
The original name of this variable was somewhat similar to this, but my goal was to make clear that this variable is only being used in the non-lazy case.